eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java

```diff
@@ -0,0 +1,551 @@
+import org.eclipse.collections.impl.block.factory.Predicates;
+import org.eclipse.collections.impl.block.procedure.MutatingAggregationProcedure;
+import org.eclipse.collections.impl.block.procedure.NonMutatingAggregationProcedure;
+import org.eclipse.collections.impl.factory.Bags;
+import org.eclipse.collections.impl.factory.Stacks;
+import org.eclipse.collections.impl.factory.primitive.BooleanStacks;
+import org.eclipse.collections.impl.factory.primitive.ByteStacks;
+import org.eclipse.collections.impl.factory.primitive.CharStacks;
+import org.eclipse.collections.impl.factory.primitive.DoubleStacks;
+import org.eclipse.collections.impl.factory.primitive.FloatStacks;
+import org.eclipse.collections.impl.factory.primitive.IntStacks;
+import org.eclipse.collections.impl.factory.primitive.LongStacks;
+import org.eclipse.collections.impl.factory.primitive.ShortStacks;
+import org.eclipse.collections.impl.list.mutable.FastList;
+import org.eclipse.collections.impl.map.mutable.UnifiedMap;
+import org.eclipse.collections.impl.multimap.list.FastListMultimap;
+import org.eclipse.collections.impl.partition.stack.PartitionArrayStack;
+import org.eclipse.collections.impl.partition.stack.PartitionArrayStack.PartitionPredicate2Procedure;
+import org.eclipse.collections.impl.partition.stack.PartitionArrayStack.PartitionProcedure;
+import org.eclipse.collections.impl.stack.mutable.ArrayStack;
+
+final class ImmutableNotEmptyStack<T>
+        extends AbstractRichIterable<T>
+        implements ImmutableStack<T>, Serializable
+{
+    private static final long serialVersionUID = 1L;
+    private final T element;
+    private final ImmutableStack<T> next;
+    private final int size;
+
+    ImmutableNotEmptyStack(T element, ImmutableStack<T> next)
+    {
+        this.element = element;
+        this.next = Objects.requireNonNull(next);
+        this.size = next.size() + 1;
+    }
+
+    @Override
+    public ImmutableStack<T> push(T item)
+    {
+        return new ImmutableNotEmptyStack<T>(item, this);
+    }
+
+    @Override
+    public ImmutableStack<T> pop()
+    {
+        return this.next;
+    }
+
+    @Override
+    public ImmutableStack<T> pop(int count)
+    {
+        this.checkNegativeCount(count);
+        if (this.checkZeroCount(count))
+        {
+            return this;
+        }
+        this.checkSizeLessThanCount(count);
+
+        ImmutableStack<T> pointer = this;
+        for (int i = 0; i < count; i++)
+        {
+            pointer = pointer.pop();
+        }
+        return pointer;
+    }
+
+    @Override
+    public T peek()
+    {
+        return this.element;
+    }
+
+    @Override
+    public ListIterable<T> peek(int count)
+    {
+        this.checkNegativeCount(count);
+        if (this.checkZeroCount(count))
+        {
+            return FastList.newList();
+        }
+        this.checkSizeLessThanCount(count);
+        return FastList.newList(this.asLazy().take(count));
```

Comment:
Can this be `this.asLazy().take(count).toList()`?

Reply:
Yes, done